packages/@lwc/template-compiler/src/__tests__/fixtures/directive-lwc-if-else/interop/foreach-if/actual.html

```diff
@@ -0,0 +1,4 @@
+<template>
+    <template for:each={items} for:item="item" lwc:if={visible}>Conditional Iteration</template>
+    <template for:each={altItems} for:item="item" lwc:else>Else Iteration</template>
```

Comment:
I would like to see is this example:

```html

    <template for:each={items} for:item="item">
        <template lwc:if={visible}>
            <div key={item.key}>Conditional Iteration</div>
        </template>
    </template>
```

I suspect it will generate an iteration with every element being the if, and thus having the same key. it is not ideal, but it will work. a few things to notice/investigate there:
- the "move element", will never happen. side note: the diffing algo does not know how to "move" a fragment.
- if you insert an element at the beginning, all of the elements in the foreach will be re-created. (maybe?)

with that in mind, I wonder, is there something that we can do to improve/avoid this use case? what's your thought about it?

Reply:
That's an interesting question you bring up.... Here are my initial thoughts about this:

This does ends up being generated for your example:
```javascript
return api_iterator($cmp.items, function (item) {
    return $cmp.visible
      ? api_fragment(
          0,
          [
            api_element(
              "div",
              {
                key: api_key(1, item.key),
              },
              [api_text("Conditional Iteration")]
            ),
          ],
          0
        )
      : null;
  });
```

### Constant Conditional in Iteration
For a simple example like the one above, where the conditional isn't dependent on the list, I would think we can avoid this use case:
```html
<template lwc:if={visible}>
    <template for:each={items} for:item="item">
        <div key={item.key}>Conditional Iteration</div>
    </template>
</template>
```
No different from general programming:
```javascript
for(let i=0; i<items.length; i++) {
    if(visible) { ... };
}
```
versus
```javascript
if (visible) {
    for(let i=0; i<items.length; i++) {
        ...
    }
}
```
Given the perf benefits from the diffing algo, we could call that out in our docs as a best practice.

### Dynamic Conditional
Now, what if our condition is based on the items in the iteration?
```html
<template for:each={items} for:item="item">
    <template lwc:if={item.visible}>
        <div key={item.key}>{item.value}</div>
    </template>
    <template lwc:else>
        <div>Placeholder</div>
    </template>
</template>
```
We could rewrite this as
```html
<template for:each={items} for:item="item">
    <div lwc:if={item.visible} key={item.key}>{item.value}</div>
    <div lwc:else key={item.key}>Placeholder</div>
</template>
```
In this situation, we're forced to give a key for the DIVs.

Maybe the user-provided key should be used for the fragment then:
```javascript
item.visible ?
    api_fragment( api_key(N, item.key), children) :
    api_fragment( api_key(N, item.key), elseChildren)
```
with a warning if `else.key !== if.key`. That's not an easy change to make though, because currently we don't get the `key` attribute until after we're done processing the if-directives.

If we're doing this for elements and components, maybe we also need to loosen the restriction on using the `key` attribute with `<template>` so users can do the following:
```html
<template for:each={items} for:item="item">
    <template lwc:if={item.visible} key={item.key}>
        <div>{item.value}</div>
    </template>
    <template lwc:else key={item.key}>
        <div>Placeholder</div>
    </template>
</template>
```
I don't like that approach though because that implies we're rendering the templates, but that seems like the only way to correctly get the `item.key` onto the if node.

Alternatively, there's also the proposal here https://github.com/salesforce/lwc/issues/2697 that I found interesting. I think that would resolve most of the issues we have here.

I've verified that it does work, just less optimally, and I added a few tests. What do you think about the proposal in 2697, or allowing for ` api_fragment( api_key(N, item.key), children) `?

Reply:
> Constant Conditional in Iteration

nice, I didn't think about this case... it makes sense, and like you mention we should put it in the docs. also, this is an improvement that we can do in the compiler in a followup pr.

> Dynamic Conditional

>> with a warning if else.key !== if.key. That's not an easy change to make though, because currently we don't get the key attribute until after we're done processing the if-directives.

correct, but I don't think we need it, as long as the iteration items have a key, we should be good about it, it will be less optimal, but it is in the author's hands.

>> Alternatively, there's also the proposal here https://github.com/salesforce/lwc/issues/2697 that I found interesting. I think that would resolve most of the issues we have here.

iirc, #2957 implements something similar and passes down the key of the foreach to the if (if:true), we could do something like that, if we want to go with that approach. Notice that for this approach, you will need to modify the diffing algo and provide a logic to "move" fragments (we can talk about this one, but it also has it's impact).

Alternatively, what do you (and the team) think about a  more radical approach? We could not allow the `if` directive to be a direct child of the `for:each` directive, it must be a descendant of a keyed element. With that constraint, all of the above issues will be solved.

Reply:
> Dynamic Conditional
> > with a warning if else.key !== if.key. That's not an easy change to make though, because currently we don't get the key attribute until after we're done processing the if-directives.
> > > correct, but I don't think we need it, as long as the iteration items have a key, we should be good about it, it will be less optimal, but it is in the author's hands.

Just to clarify, right now without any changes,
```html
<template for:each={items} for:item="item">
    <div lwc:if={item.visible} key={item.key}>{item.value}</div>
</template>
```
gets compiled as something like
```javascript
return api_iterator($cmp.items, function (item) {
    return item.visible
      ? api_fragment(
          0,
          [
            api_element(
              "div",
              {
                key: api_key(1, item.key),
              },
              [item.value]
            ),
          ],
          0
        )
      : null;
  });
```
So if we want to use the iteration items' key, we do need to make some change to support it. As-is, it works, but will be a performance hit even though it doesn't look like it from the template.

> passes down the key of the foreach to the if (if:true), we could do something like that, if we want to go with that approach. Notice that for this approach, you will need to modify the diffing algo and provide a logic to "move" fragments

That sounds like something we could consider. We haven't done the work to move for-each over to fragments yet right? I get the feeling depending on how we do that, we might address the issue here, or open a new path forward.

> We could not allow the if directive to be a direct child of the for:each directive, it must be a descendant of a keyed element.

If we disallow this, what would be the suggested approach for a user who wants to implement something like
```html
<template for:each={items} for:item="item">
    <div lwc:if={item.visible} key={item.key} class="listItem">{item.value}</div>
</template>
```
? The desired result is a series of `<div>`s that correspond to the visible items in the list:
`# of <div> === # of items where item.visible === true`.

But, if we can't have an `if` directive on the direct child of the for:each, then each direct child element will be rendered regardless of the condition, and we would have:
`# of <div> === items.length`, with `item.visible` only affecting whether the `<div>` elements are empty.

At least this morning, I can't think of a way to support this use case if we disallow the `if` directive on the direct child. Do you have any thoughts?

Reply:
```html
<template for:each={items} for:item="item">
    <div lwc:if={item.visible} key={item.key} class="listItem">{item.value}</div>
</template>
```
yeah, it's not ideal. to implement that, I would do a getter/iterator that filters out the items that are not visible.

---------

>> passes down the key of the foreach to the if (if:true), we could do something like that, if we want to go with that approach. Notice that for this approach, you will need to modify the diffing algo and provide a logic to "move" fragments

> That sounds like something we could consider. We haven't done the work to move for-each over to fragments yet right? I get the feeling depending on how we do that, we might address the issue here, or open a new path forward.

We haven't done the work to move for-each over to fragments. atm we are not considering it because of the backward compatibility issues(observable changes). but I think some work could be done when the `if` directive is a direct child of the `for` (if the template is using the `if` directive, then is new)